pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go

```diff
@@ -339,8 +339,14 @@ func (c *clusterAutoscaler) emptyManagedResourceSecret() *corev1.Secret {
 
 func (c *clusterAutoscaler) computeCommand() []string {
 	var (
-		scaleDownUnneededTime  = metav1.Duration{Duration: 30 * time.Minute}
-		scaleDownDelayAfterAdd = metav1.Duration{Duration: time.Hour}
+		expander                      = "least-waste"
+		scaleDownUtilizationThreshold = 0.5
+		maxNodeProvisionTime          = metav1.Duration{Duration: 20 * time.Minute}
+		scaleDownUnneededTime         = metav1.Duration{Duration: 30 * time.Minute}
+		scaleDownDelayAfterAdd        = metav1.Duration{Duration: time.Hour}
+		scaleDownDelayAfterFailure    = metav1.Duration{Duration: 3 * time.Minute}
+		scaleDownDelayAfterDelete     = metav1.Duration{Duration: 0}
+		scanInterval                  = metav1.Duration{Duration: 10 * time.Second}
```

Comment:
I am afraid these are the actual default values and they can easily changed without doing the same in the API. I would propose to always use the one from the API as all of them are now defaulted.

Reply:
From my local tests, I see that the defaulting isn't done for existing shoot objects. So not sure what would be the best way for these objects. As an improvement I have updated this logic to [make use of the defaulting method here](https://github.com/gardener/gardener/blob/ed626583966d98e816e697c1b4780373f0a3739d/pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go#L339-L376). Please do suggest if there is a better way.

Reply:
I don't have better idea. I just checked other components and they also duplicate the default values, but I really think this is error prone and the approach that you now implemented seems better to me.


The defaulting does not happen because of
```golang
	if in.Spec.Kubernetes.ClusterAutoscaler != nil {
		SetDefaults_ClusterAutoscaler(in.Spec.Kubernetes.ClusterAutoscaler)
	}
```

in the `zz_generated.defaults.go`. You see, it expects the `ClusterAutoscaler` to be not nil in order to default the values, but with 
https://github.com/gardener/gardener/blob/6cdb8dc690c2464558c228105852eb3ac7c2c05a/pkg/apis/core/v1beta1/types_shoot.go#L353-L354

it is `nil` when skipped.

Reply:
Oh okay. Thanks for the detailed explanation. I knew something did not default properly, however wasn't sure where. This explanation definitely helps. Thanks.